File: app/models/cms.py
"""
CMS Models for D2C Storefront Content Management

Content types:
- CMSBanner: Hero banners with images, CTAs, scheduling
- CMSUsp: USPs/Features with icons
- CMSTestimonial: Customer testimonials
- CMSAnnouncement: Announcement bar messages
- CMSPage: Static pages with rich text content
- CMSPageVersion: Version history for pages
- CMSSeo: SEO settings per page
"""

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import TYPE_CHECKING, Optional, List

from sqlalchemy import String, Boolean, DateTime, ForeignKey, Integer, Text, CheckConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, JSONB

from app.database import Base

if TYPE_CHECKING:
    from app.models.user import User


# ==================== Enums (stored as VARCHAR in DB) ====================

class CMSPageStatus(str, Enum):
    """Page publish status"""
    DRAFT = "DRAFT"
    PUBLISHED = "PUBLISHED"
    ARCHIVED = "ARCHIVED"


class CMSAnnouncementType(str, Enum):
    """Announcement bar type/style"""
    INFO = "INFO"
    WARNING = "WARNING"
    PROMO = "PROMO"
    SUCCESS = "SUCCESS"


# ==================== CMS Banner Model ====================

class CMSBanner(Base):
    """
    Hero banner for D2C storefront homepage.
    Supports scheduling, multiple banners with sort order.
    """
    __tablename__ = "cms_banners"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Content
    title: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        comment="Banner headline"
    )
    subtitle: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Banner subheadline/description"
    )
    image_url: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        comment="Full-size banner image URL"
    )
    thumbnail_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Thumbnail image for admin preview"
    )
    mobile_image_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Mobile-optimized image URL"
    )

    # CTA (Call to Action)
    cta_text: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        comment="Button text e.g., 'Shop Now'"
    )
    cta_link: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Button link URL"
    )

    # Positioning
    text_position: Mapped[str] = mapped_column(
        String(20),
        default="left",
        nullable=False,
        comment="Text alignment: left, center, right"
    )
    text_color: Mapped[str] = mapped_column(
        String(20),
        default="white",
        nullable=False,
        comment="Text color: white, dark"
    )

    # Display settings
    sort_order: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False,
        comment="Display order (lower = first)"
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False
    )

    # Scheduling
    starts_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Start showing banner at this time"
    )
    ends_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Stop showing banner after this time"
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    # Relationships
    creator: Mapped[Optional["User"]] = relationship("User", foreign_keys=[created_by])

    def __repr__(self) -> str:
        return f"<CMSBanner(title='{self.title}', active={self.is_active})>"


# ==================== CMS USP Model ====================

class CMSUsp(Base):
    """
    USPs (Unique Selling Points) / Features for homepage.
    E.g., "Free Installation", "2 Year Warranty", etc.
    """
    __tablename__ = "cms_usps"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Content
    title: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        comment="USP headline"
    )
    description: Mapped[Optional[str]] = mapped_column(
        String(300),
        nullable=True,
        comment="Short description"
    )
    icon: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="Lucide icon name e.g., 'truck', 'shield-check'"
    )
    icon_color: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        comment="Icon color class e.g., 'text-blue-500'"
    )

    # Link (optional)
    link_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    link_text: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True
    )

    # Display
    sort_order: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    def __repr__(self) -> str:
        return f"<CMSUsp(title='{self.title}', icon='{self.icon}')>"


# ==================== CMS Testimonial Model ====================

class CMSTestimonial(Base):
    """
    Customer testimonials for homepage.
    """
    __tablename__ = "cms_testimonials"
    __table_args__ = (
        CheckConstraint('rating >= 1 AND rating <= 5', name='check_testimonial_rating'),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Customer info
    customer_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False
    )
    customer_location: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        comment="City/State"
    )
    customer_avatar_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    customer_designation: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        comment="Job title or role (optional)"
    )

    # Review content
    rating: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        comment="Rating from 1-5 stars"
    )
    content: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        comment="Testimonial text"
    )
    title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True,
        comment="Review title/headline"
    )

    # Product reference (optional)
    product_name: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True,
        comment="Product being reviewed"
    )
    product_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("products.id", ondelete="SET NULL"),
        nullable=True
    )

    # Display
    sort_order: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False
    )
    is_featured: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
        comment="Featured testimonials shown first"
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    def __repr__(self) -> str:
        return f"<CMSTestimonial(customer='{self.customer_name}', rating={self.rating})>"


# ==================== CMS Announcement Model ====================

class CMSAnnouncement(Base):
    """
    Announcement bar messages for storefront header.
    E.g., "Free shipping on orders above ₹999!"
    """
    __tablename__ = "cms_announcements"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Content
    text: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        comment="Announcement message"
    )
    link_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Link when clicked"
    )
    link_text: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        comment="Link text e.g., 'Learn More'"
    )

    # Style
    announcement_type: Mapped[str] = mapped_column(
        String(20),
        default="INFO",
        nullable=False,
        comment="INFO, WARNING, PROMO, SUCCESS"
    )
    background_color: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        comment="Custom background color"
    )
    text_color: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        comment="Custom text color"
    )

    # Scheduling
    starts_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True
    )
    ends_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True
    )

    # Display
    sort_order: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False
    )
    is_dismissible: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
        comment="Can user close this announcement"
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    def __repr__(self) -> str:
        return f"<CMSAnnouncement(text='{self.text[:50]}...', type={self.announcement_type})>"


# ==================== CMS Page Model ====================

class CMSPage(Base):
    """
    Static pages with rich text content.
    E.g., About Us, Privacy Policy, Terms & Conditions.
    """
    __tablename__ = "cms_pages"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Identification
    title: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        comment="Page title"
    )
    slug: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        unique=True,
        index=True,
        comment="URL slug e.g., 'about-us'"
    )

    # Content
    content: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="Rich text HTML content"
    )
    excerpt: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Short summary for listings"
    )

    # SEO
    meta_title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True,
        comment="SEO meta title"
    )
    meta_description: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="SEO meta description"
    )
    meta_keywords: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="SEO keywords (comma-separated)"
    )
    og_image_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Open Graph image URL"
    )
    canonical_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Canonical URL if different"
    )

    # Status
    status: Mapped[str] = mapped_column(
        String(20),
        default="DRAFT",
        nullable=False,
        comment="DRAFT, PUBLISHED, ARCHIVED"
    )
    published_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="When page was published"
    )

    # Template
    template: Mapped[str] = mapped_column(
        String(50),
        default="default",
        nullable=False,
        comment="Page template: default, full-width, landing"
    )

    # Navigation
    show_in_footer: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False
    )
    show_in_header: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False
    )
    sort_order: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )
    updated_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    # Relationships
    versions: Mapped[List["CMSPageVersion"]] = relationship(
        "CMSPageVersion",
        back_populates="page",
        cascade="all, delete-orphan",
        order_by="CMSPageVersion.version_number.desc()"
    )
    creator: Mapped[Optional["User"]] = relationship("User", foreign_keys=[created_by])
    updater: Mapped[Optional["User"]] = relationship("User", foreign_keys=[updated_by])

    def __repr__(self) -> str:
        return f"<CMSPage(title='{self.title}', slug='{self.slug}', status={self.status})>"


# ==================== CMS Page Version Model ====================

class CMSPageVersion(Base):
    """
    Version history for CMS pages.
    Created automatically when a page is updated.
    """
    __tablename__ = "cms_page_versions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # Page reference
    page_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("cms_pages.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    version_number: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        comment="Sequential version number"
    )

    # Snapshot of content at this version
    title: Mapped[str] = mapped_column(
        String(200),
        nullable=False
    )
    content: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True
    )
    meta_title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True
    )
    meta_description: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )

    # Change metadata
    change_summary: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
        comment="Description of changes in this version"
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    # Relationships
    page: Mapped["CMSPage"] = relationship("CMSPage", back_populates="versions")
    creator: Mapped[Optional["User"]] = relationship("User", foreign_keys=[created_by])

    def __repr__(self) -> str:
        return f"<CMSPageVersion(page_id={self.page_id}, version={self.version_number})>"


# ==================== CMS SEO Settings Model ====================

class CMSSeo(Base):
    """
    SEO settings for specific routes/pages.
    Allows overriding default SEO for any URL path.
    """
    __tablename__ = "cms_seo"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4
    )

    # URL/Route
    url_path: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        unique=True,
        index=True,
        comment="URL path e.g., '/', '/products', '/products/aqua-ro'"
    )

    # SEO fields
    meta_title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True
    )
    meta_description: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    meta_keywords: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    og_title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True
    )
    og_description: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    og_image_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )
    og_type: Mapped[str] = mapped_column(
        String(50),
        default="website",
        nullable=False
    )
    canonical_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True
    )

    # Robots
    robots_index: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
        comment="Allow search engine indexing"
    )
    robots_follow: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
        comment="Allow following links"
    )

    # Structured data (JSON-LD)
    structured_data: Mapped[Optional[dict]] = mapped_column(
        JSONB,
        nullable=True,
        comment="JSON-LD structured data"
    )

    # Audit
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )

    def __repr__(self) -> str:
        return f"<CMSSeo(url_path='{self.url_path}')>"

File: frontend/src/lib/api/cms.ts
/**
 * CMS API Client
 * Handles all CMS-related API calls for the dashboard
 */

import { apiClient } from './client';

// ==================== Types ====================

export interface CMSBanner {
  id: string;
  title: string;
  subtitle?: string;
  image_url: string;
  thumbnail_url?: string;
  mobile_image_url?: string;
  cta_text?: string;
  cta_link?: string;
  text_position: 'left' | 'center' | 'right';
  text_color: 'white' | 'dark';
  sort_order: number;
  is_active: boolean;
  starts_at?: string;
  ends_at?: string;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CMSBannerCreate {
  title: string;
  subtitle?: string;
  image_url: string;
  thumbnail_url?: string;
  mobile_image_url?: string;
  cta_text?: string;
  cta_link?: string;
  text_position?: 'left' | 'center' | 'right';
  text_color?: 'white' | 'dark';
  sort_order?: number;
  is_active?: boolean;
  starts_at?: string;
  ends_at?: string;
}

export interface CMSUsp {
  id: string;
  title: string;
  description?: string;
  icon: string;
  icon_color?: string;
  link_url?: string;
  link_text?: string;
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CMSUspCreate {
  title: string;
  description?: string;
  icon: string;
  icon_color?: string;
  link_url?: string;
  link_text?: string;
  sort_order?: number;
  is_active?: boolean;
}

export interface CMSTestimonial {
  id: string;
  customer_name: string;
  customer_location?: string;
  customer_avatar_url?: string;
  customer_designation?: string;
  rating: number;
  content: string;
  title?: string;
  product_name?: string;
  product_id?: string;
  sort_order: number;
  is_featured: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CMSTestimonialCreate {
  customer_name: string;
  customer_location?: string;
  customer_avatar_url?: string;
  customer_designation?: string;
  rating: number;
  content: string;
  title?: string;
  product_name?: string;
  product_id?: string;
  sort_order?: number;
  is_featured?: boolean;
  is_active?: boolean;
}

export interface CMSAnnouncement {
  id: string;
  text: string;
  link_url?: string;
  link_text?: string;
  announcement_type: 'INFO' | 'WARNING' | 'PROMO' | 'SUCCESS';
  background_color?: string;
  text_color?: string;
  starts_at?: string;
  ends_at?: string;
  sort_order: number;
  is_dismissible: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CMSAnnouncementCreate {
  text: string;
  link_url?: string;
  link_text?: string;
  announcement_type?: 'INFO' | 'WARNING' | 'PROMO' | 'SUCCESS';
  background_color?: string;
  text_color?: string;
  starts_at?: string;
  ends_at?: string;
  sort_order?: number;
  is_dismissible?: boolean;
  is_active?: boolean;
}

export interface CMSPageVersion {
  id: string;
  page_id: string;
  version_number: number;
  title: string;
  content?: string;
  meta_title?: string;
  meta_description?: string;
  change_summary?: string;
  created_at: string;
  created_by?: string;
}

export interface CMSPage {
  id: string;
  title: string;
  slug: string;
  content?: string;
  excerpt?: string;
  meta_title?: string;
  meta_description?: string;
  meta_keywords?: string;
  og_image_url?: string;
  canonical_url?: string;
  status: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  published_at?: string;
  template: 'default' | 'full-width' | 'landing';
  show_in_footer: boolean;
  show_in_header: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
  versions: CMSPageVersion[];
}

export interface CMSPageBrief {
  id: string;
  title: string;
  slug: string;
  status: string;
  published_at?: string;
  updated_at: string;
}

export interface CMSPageCreate {
  title: string;
  slug: string;
  content?: string;
  excerpt?: string;
  meta_title?: string;
  meta_description?: string;
  meta_keywords?: string;
  og_image_url?: string;
  canonical_url?: string;
  status?: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  template?: 'default' | 'full-width' | 'landing';
  show_in_footer?: boolean;
  show_in_header?: boolean;
  sort_order?: number;
}

export interface CMSSeo {
  id: string;
  url_path: string;
  meta_title?: string;
  meta_description?: string;
  meta_keywords?: string;
  og_title?: string;
  og_description?: string;
  og_image_url?: string;
  og_type: string;
  canonical_url?: string;
  robots_index: boolean;
  robots_follow: boolean;
  structured_data?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CMSSeoCreate {
  url_path: string;
  meta_title?: string;
  meta_description?: string;
  meta_keywords?: string;
  og_title?: string;
  og_description?: string;
  og_image_url?: string;
  og_type?: string;
  canonical_url?: string;
  robots_index?: boolean;
  robots_follow?: boolean;
  structured_data?: Record<string, unknown>;
}

// Site Settings
export interface CMSSiteSetting {
  id: string;
  setting_key: string;
  setting_value?: string;
  setting_type: 'text' | 'textarea' | 'url' | 'boolean' | 'number' | 'image';
  setting_group: string;
  label?: string;
  description?: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface CMSSiteSettingCreate {
  setting_key: string;
  setting_value?: string;
  setting_type: 'text' | 'textarea' | 'url' | 'boolean' | 'number' | 'image';
  setting_group: string;
  label?: string;
  description?: string;
  sort_order?: number;
}

// Menu Items
export interface CMSMenuItem {
  id: string;
  menu_location: 'header' | 'footer_quick' | 'footer_service';
  title: string;
  url: string;
  icon?: string;
  target: '_self' | '_blank';
  parent_id?: string;
  sort_order: number;
  is_active: boolean;
  show_on_mobile: boolean;
  css_class?: string;
  created_at: string;
  updated_at: string;
  children?: CMSMenuItem[];
}

export interface CMSMenuItemCreate {
  menu_location: 'header' | 'footer_quick' | 'footer_service';
  title: string;
  url: string;
  icon?: string;
  target?: '_self' | '_blank';
  parent_id?: string;
  sort_order?: number;
  is_active?: boolean;
  show_on_mobile?: boolean;
  css_class?: string;
}

// Feature Bars
export interface CMSFeatureBar {
  id: string;
  icon: string;
  title: string;
  subtitle?: string;
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CMSFeatureBarCreate {
  icon: string;
  title: string;
  subtitle?: string;
  sort_order?: number;
  is_active?: boolean;
}

export interface ListResponse<T> {
  items: T[];
  total: number;
}

// ==================== API Functions ====================

export const cmsApi = {
  // Banners
  banners: {
    list: (params?: { is_active?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSBanner>>('/cms/banners', { params }),

    get: (id: string) =>
      apiClient.get<CMSBanner>(`/cms/banners/${id}`),

    create: (data: CMSBannerCreate) =>
      apiClient.post<CMSBanner>('/cms/banners', data),

    update: (id: string, data: Partial<CMSBannerCreate>) =>
      apiClient.put<CMSBanner>(`/cms/banners/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/banners/${id}`),

    reorder: (ids: string[]) =>
      apiClient.put<CMSBanner[]>('/cms/banners/reorder', { ids }),
  },

  // USPs
  usps: {
    list: (params?: { is_active?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSUsp>>('/cms/usps', { params }),

    get: (id: string) =>
      apiClient.get<CMSUsp>(`/cms/usps/${id}`),

    create: (data: CMSUspCreate) =>
      apiClient.post<CMSUsp>('/cms/usps', data),

    update: (id: string, data: Partial<CMSUspCreate>) =>
      apiClient.put<CMSUsp>(`/cms/usps/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/usps/${id}`),

    reorder: (ids: string[]) =>
      apiClient.put<CMSUsp[]>('/cms/usps/reorder', { ids }),
  },

  // Testimonials
  testimonials: {
    list: (params?: { is_active?: boolean; is_featured?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSTestimonial>>('/cms/testimonials', { params }),

    get: (id: string) =>
      apiClient.get<CMSTestimonial>(`/cms/testimonials/${id}`),

    create: (data: CMSTestimonialCreate) =>
      apiClient.post<CMSTestimonial>('/cms/testimonials', data),

    update: (id: string, data: Partial<CMSTestimonialCreate>) =>
      apiClient.put<CMSTestimonial>(`/cms/testimonials/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/testimonials/${id}`),
  },

  // Announcements
  announcements: {
    list: (params?: { is_active?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSAnnouncement>>('/cms/announcements', { params }),

    get: (id: string) =>
      apiClient.get<CMSAnnouncement>(`/cms/announcements/${id}`),

    create: (data: CMSAnnouncementCreate) =>
      apiClient.post<CMSAnnouncement>('/cms/announcements', data),

    update: (id: string, data: Partial<CMSAnnouncementCreate>) =>
      apiClient.put<CMSAnnouncement>(`/cms/announcements/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/announcements/${id}`),
  },

  // Pages
  pages: {
    list: (params?: { status?: string; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSPageBrief>>('/cms/pages', { params }),

    get: (id: string) =>
      apiClient.get<CMSPage>(`/cms/pages/${id}`),

    create: (data: CMSPageCreate) =>
      apiClient.post<CMSPage>('/cms/pages', data),

    update: (id: string, data: Partial<CMSPageCreate>) =>
      apiClient.put<CMSPage>(`/cms/pages/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/pages/${id}`),

    publish: (id: string) =>
      apiClient.post<CMSPage>(`/cms/pages/${id}/publish`),

    getVersions: (id: string) =>
      apiClient.get<CMSPageVersion[]>(`/cms/pages/${id}/versions`),

    revertToVersion: (id: string, versionNumber: number) =>
      apiClient.post<CMSPage>(`/cms/pages/${id}/revert/${versionNumber}`),
  },

  // SEO
  seo: {
    list: (params?: { skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSSeo>>('/cms/seo', { params }),

    get: (id: string) =>
      apiClient.get<CMSSeo>(`/cms/seo/${id}`),

    create: (data: CMSSeoCreate) =>
      apiClient.post<CMSSeo>('/cms/seo', data),

    update: (id: string, data: Partial<CMSSeoCreate>) =>
      apiClient.put<CMSSeo>(`/cms/seo/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/seo/${id}`),
  },

  // Site Settings
  settings: {
    list: (params?: { group?: string; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSSiteSetting>>('/cms/settings', { params }),

    get: (key: string) =>
      apiClient.get<CMSSiteSetting>(`/cms/settings/${key}`),

    create: (data: CMSSiteSettingCreate) =>
      apiClient.post<CMSSiteSetting>('/cms/settings', data),

    update: (key: string, data: Partial<CMSSiteSettingCreate>) =>
      apiClient.put<CMSSiteSetting>(`/cms/settings/${key}`, data),

    bulkUpdate: (settings: Record<string, string>) =>
      apiClient.put<CMSSiteSetting[]>('/cms/settings-bulk', { settings }),

    delete: (key: string) =>
      apiClient.delete(`/cms/settings/${key}`),
  },

  // Menu Items
  menuItems: {
    list: (params?: { location?: string; is_active?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSMenuItem>>('/cms/menu-items', { params }),

    get: (id: string) =>
      apiClient.get<CMSMenuItem>(`/cms/menu-items/${id}`),

    create: (data: CMSMenuItemCreate) =>
      apiClient.post<CMSMenuItem>('/cms/menu-items', data),

    update: (id: string, data: Partial<CMSMenuItemCreate>) =>
      apiClient.put<CMSMenuItem>(`/cms/menu-items/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/menu-items/${id}`),

    reorder: (ids: string[]) =>
      apiClient.put<CMSMenuItem[]>('/cms/menu-items/reorder', { ids }),
  },

  // Feature Bars
  featureBars: {
    list: (params?: { is_active?: boolean; skip?: number; limit?: number }) =>
      apiClient.get<ListResponse<CMSFeatureBar>>('/cms/feature-bars', { params }),

    get: (id: string) =>
      apiClient.get<CMSFeatureBar>(`/cms/feature-bars/${id}`),

    create: (data: CMSFeatureBarCreate) =>
      apiClient.post<CMSFeatureBar>('/cms/feature-bars', data),

    update: (id: string, data: Partial<CMSFeatureBarCreate>) =>
      apiClient.put<CMSFeatureBar>(`/cms/feature-bars/${id}`, data),

    delete: (id: string) =>
      apiClient.delete(`/cms/feature-bars/${id}`),

    reorder: (ids: string[]) =>
      apiClient.put<CMSFeatureBar[]>('/cms/feature-bars/reorder', { ids }),
  },
};

export default cmsApi;
